slides/week3/git.org

```diff
@@ -0,0 +1,200 @@
+    # git branch <BRANCH_NAME>
+
+    # Create two branches, starting from the current commit
+
+    git branch stable
+    git branch beta
+  #+END_SRC
+
+** Switching Branches
+
+- only one branch can be 'worked on' at a time.
+- ~git status~ will display info on your current branch
+- ~git checkout~ let's you switch between branches
+- Let's checkout to the *stable* branch now
+  #+BEGIN_SRC sh
+    git status
+    #> On branch master
+    #> nothing to commit, working tree clean
+
+    # git checkout <BRANCH_NAME>
+
+    git checkout stable
+    #> Switched to branch 'stable
+
+    git status
+    #> On branch stable
+    #> nothing to commit, working tree clean
+  #+END_SRC
+
+** First Stable Commit
+- Let's add a commit to the stable branch
+  #+BEGIN_SRC sh
+    echo "Stability counts!" > stable_release.txt
+    git add -A
+    git commit -m "Add stable_release.txt"
+  #+END_SRC
+- This will *not affect* any other branches
+- The other branches will stay behind on 'initial commit'
+  #+BEGIN_SRC
+    5592388 * stable      Add stable_release.txt
+    844a162 * master beta Initial commit
+  #+END_SRC
+
+** First Beta Commit
+- Let's make a experimental commit on the beta branch
+  #+BEGIN_SRC sh
+    git checkout beta
+    echo "This feature is unstable!" > beta_release.txt
+    git add -A
+    git commit -m "Add beta_release.txt"
+  #+END_SRC
+- Since the beta branch was *behind* the stable branch, this creates a 'fork' in the commit history
+- Notice, master has not moved at all (since no commits have been made on it)
+  #+BEGIN_SRC
+    1c65f7b * beta Add beta_release.txt
+    5592388 | * stable Add stable_release.txt
+            |/
+    844a162 * master Initial commit
+  #+END_SRC
+- Play around, and checkout to the various branches!
+  #+BEGIN_SRC shell
+    git checkout <BRANCH>
+    ls
+  #+END_SRC
+
+** Let's make beta official
+- Let's bring the hard work from the beta branch onto the stable branch
+- Right now, the stable and the beta branch have 'forked', we need to bring them back together
+- This is a *merge*
+  #+BEGIN_SRC sh
+    # Checkout to the branch we want to merge **into** aka the "base branch"
+    git checkout stable
+    # Merge the branch we want (beta) into the current branch (stable)
+    git merge beta
+    # This will launch an editor, save and quit it to complete
+  #+END_SRC
+- This does not change any other branch
+  #+BEGIN_SRC
+    ddc2df9 *   stable Merge branch 'beta' into stable
+            |\
+    1c65f7b | * beta Add beta_release.txt
+    5592388 * | Add stable_release.txt
+            |/
+    844a162 * master Initial commit
+  #+END_SRC
+
+** Remotes vs Branches vs Forks
+- A fork or close does a full copy of the git repo
+- A remote is a pointer to another copy of this git repo
+- A repository is a collection of *branches*
+- You can only directly modify local branches, but:
+  - You can merge from a remote into your branch (pull)
+  - You can merge from your branch to a remote (push)
```

Comment:
this is still a merge (plus a fetch). You can even use the merge command to do it (with a fetch).

I want to make an explicit point here, how branches on remotes are no different from branches locally, and the same terminology and logic can be used. If remotes are treated specially it makes understanding git harder, at least in my opinion. This was one of the confusion points when I learnt git.

```
git fetch origin
git merge origin/master
```